### PR TITLE
Interceptor framework refactoring

### PR DESCRIPTION
**Pull request by jamesagnew**
* Move Interceptor framework from hapi-fhir-jpaserver-model into hapi-fhir-base and remove all dependency on Spring
* Sync the latest validator and FHIRPath engines from the core library and deal with all kinds of fallout from that
* A couple of bug fixes needed by the EMPI branch in CDR

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseHapiTerminologySvcImpl.java [hunk 1]
```diff
@@ -1405,6 +1405,8 @@ public boolean supportsSystem(String theSystem) {
 					targets.add(scrollableResultsIterator.next());
 				}
 
+				scrollableResults.close();
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseHapiTerminologySvcImpl.java]: Why not a try with resources here?  Might not close if there's an exception...

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/util/ResourceProviderFactory.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+public class ResourceProviderFactory {
+
+	private List<Supplier<Object>> mySuppliers = new ArrayList<>();
+
+	public void addSupplier(@Nonnull Supplier<Object> theSupplier) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/util/ResourceProviderFactory.java]: No-one calls this method.  How do suppliers ever get added?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseHapiTerminologySvcImpl.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseHapiTerminologySvcImpl.java]: oh yeah, good call.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/util/ResourceProviderFactory.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/util/ResourceProviderFactory.java]: Yeah good call- I think I had a use in mind but it never happened. Removing.

**Pull request merged**